server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -217,16 +219,12 @@ synchronized boolean hasSnapshottedCommits() {
     }
 
     /**
-     * Checks if the deletion policy can release some index commits with the latest global checkpoint.
+     * Checks if the deletion policy can delete some index commits with the latest global checkpoint.
      */
     boolean hasUnreferencedCommits() throws IOException {
-        final IndexCommit lastCommit = this.lastCommit;
-        if (safeCommit != lastCommit) { // Race condition can happen but harmless
-            final long maxSeqNoFromLastCommit = Long.parseLong(lastCommit.getUserData().get(SequenceNumbers.MAX_SEQ_NO));
-            // We can clean up the current safe commit if the last commit is safe
-            return globalCheckpointSupplier.getAsLong() >= maxSeqNoFromLastCommit;
-        }
-        return false;
+        final IndexCommit nextSafeCommit = this.nextSafeCommit;
+        return nextSafeCommit != null
+            && Long.parseLong(nextSafeCommit.getUserData().get(SequenceNumbers.MAX_SEQ_NO)) <= globalCheckpointSupplier.getAsLong();
```

Comment:
Could we just store the max-seq-no of `nextSafeCommit` instead? Not a big deal, but I think this method is called more frequently than `onCommit` and thus it seems more logical to parse out the number in `onCommit`.

Reply:
Adjusted in https://github.com/elastic/elasticsearch/pull/48559/commits/d66389ab5df2b11b7a496bd8d86c3d778612a976.